lib/heex_formatter/phases/ensure_line_breaks.ex

```diff
@@ -39,15 +39,30 @@ defmodule HeexFormatter.Phases.NewLines do
   """
   def run(nodes, _opts) do
     initial_state = %{nodes: [], previous: nil, index: 0, length: length(nodes)}
-    result = Enum.reduce(nodes, initial_state, &may_add_new_line/2)
+    result = Enum.reduce(nodes, initial_state, &ensure_correct_line_breaks/2)
     result.nodes
   end
 
-  defp may_add_new_line({:text, _text, _meta} = node, acc) do
-    update_state(acc, [node], node)
+  defp ensure_correct_line_breaks({:text, text, meta} = node, acc) do
+    cond do
+      # In case it is a line break and it is the last interaction, we want to keep
+      # this line break in the end of the nodes.
+      line_break?(node) and last_interaction?(acc) ->
+        update_state(acc, [{:text, "\n", meta}], acc.previous)
+
+      # Skip this node in case this is a line break so that we will keep just one
+      # line break among tags.
+      line_break?(node) ->
+        update_state(acc, [], acc.previous)
+
+      # Here we know that it is not a line break so we can safely trim it to
+      # remove extra spaces from the string.
+      true ->
+        update_state(acc, [{:text, String.trim(text), meta}], node)
```

Comment:
We can trim the string here which makes the module `TagWhitespace` redundant for now.